src/shared/utilities/workspaceUtils.ts

```diff
@@ -71,3 +73,42 @@ export async function addFolderToWorkspace(folder: { uri: vscode.Uri; name?: str
         }
     }
 }
+
+/**
+ * Finds the closest file (specified by the search pattern) to the specified source file.
+ * Checks parent directories up until the top level workspace folder for the source file.
+ * @param sourceCodeUri Source file to look upwards from
+ * @param searchPattern File to find in same folder or parent, up until the source file's top level workspace folder.
+ * @param findWorkspaceFiles Only used for tests
+ */
+export async function findParentProjectFile(
+    sourceCodeUri: vscode.Uri,
+    searchPattern: string,
+    findWorkspaceFiles: typeof vscode.workspace.findFiles = vscode.workspace.findFiles
+): Promise<vscode.Uri | undefined> {
+    const workspaceFolder = vscode.workspace.getWorkspaceFolder(sourceCodeUri)
+    if (!workspaceFolder) {
+        return undefined
+    }
+
+    const workspaceProjectFiles: vscode.Uri[] = await findWorkspaceFiles(
+        new vscode.RelativePattern(workspaceFolder, searchPattern)
+    )
+
+    // Use the project file "closest" in the parent chain to sourceCodeUri
+    let parentProjectFiles = workspaceProjectFiles
+        .filter(uri => {
+            const dirname = dirnameWithTrailingSlash(uri.fsPath)
+
+            return sourceCodeUri.fsPath.startsWith(dirname)
+        })
+        .sort((a, b) => {
+            if (isInDirectory(path.parse(a.fsPath).dir, path.parse(b.fsPath).dir)) {
+                return 1
+            }
+
+            return -1
+        })
+
+    return parentProjectFiles.length === 0 ? undefined : parentProjectFiles[0]
```

Comment:
I think this is redundant since parentProjectFiles[0] returns undefined if it's empty.